libpod/image/image.go

```diff
@@ -149,43 +210,344 @@ func (i *Image) ID() string {
+	return err
+}
+
+func (i *Image) GetImageInspectInfo() (*inspect.ImageData, error) {
+
+	layer, err := i.imageruntime.store.Layer(i.TopLayer())
+
+	if err != nil {
+		return nil, errors.Wrapf(err, "error reading information about layer %q", i.TopLayer())
+	}
+	// TODO: I think this might the wrong way to get this? double check it
+	size, err := i.imageruntime.store.DiffSize(layer.Parent, layer.ID)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error determining size of layer %q", layer.ID)
+	}
+	driverData, err := driver.GetDriverData(i.imageruntime.store, layer.ID)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error getting graph driver info %q", i.ID())
+	}
+	return i.getImageData(size, driverData)
+}
+
+func (i *Image) getImageData(size int64, driver *inspect.Data) (*inspect.ImageData, error) {
+	imgRef, err := i.toImageRef()
+	if err != nil {
+		return nil, errors.Wrapf(err, "error reading image %q", i.ID())
+	}
+
+	imgSize, err := imgRef.Size()
+	if err != nil {
+		return nil, errors.Wrapf(err, "error reading size of image %q", i.ID())
+	}
+	manifest, manifestType, err := imgRef.Manifest()
+	if err != nil {
+		return nil, errors.Wrapf(err, "error reading manifest for image %q", i.ID())
+	}
+	imgDigest := digest.Digest("")
+	if len(manifest) > 0 {
+		imgDigest = digest.Canonical.FromBytes(manifest)
+	}
+	annotations := annotations(manifest, manifestType)
+
+	ociv1Img, err := imgRef.OCIConfig()
+	if err != nil {
+		return nil, errors.Wrapf(err, "error getting oci image info %q", i.ID())
+	}
+	info, err := imgRef.Inspect()
+	if err != nil {
+		return nil, errors.Wrapf(err, "error getting image info %q", i.ID())
+	}
+
+	var repoDigests []string
+	for _, name := range i.Names() {
+		repoDigests = append(repoDigests, strings.SplitN(name, ":", 2)[0]+"@"+imgDigest.String())
+	}
+
+	data := &inspect.ImageData{
+		ID:              i.ID(),
+		RepoTags:        i.Names(),
+		RepoDigests:     repoDigests,
+		Comment:         ociv1Img.History[0].Comment,
+		Created:         ociv1Img.Created,
+		Author:          ociv1Img.Author,
+		Architecture:    ociv1Img.Architecture,
+		Os:              ociv1Img.OS,
+		ContainerConfig: &ociv1Img.Config,
+		Version:         info.DockerVersion,
+		Size:            size,
+		VirtualSize:     size + imgSize,
+		Annotations:     annotations,
+		Digest:          imgDigest,
+		Labels:          info.Labels,
+		RootFS: &inspect.RootFS{
+			Type:   ociv1Img.RootFS.Type,
+			Layers: ociv1Img.RootFS.DiffIDs,
+		},
+		GraphDriver: driver,
+	}
+	return data, nil
+}
+
+func annotations(manifest []byte, manifestType string) map[string]string {
+	annotations := make(map[string]string)
+	switch manifestType {
+	case ociv1.MediaTypeImageManifest:
+		var m ociv1.Manifest
+		if err := json.Unmarshal(manifest, &m); err == nil {
+			for k, v := range m.Annotations {
+				annotations[k] = v
+			}
+		}
+	}
+	return annotations
+}
+
+func (i *Image) toImageRef() (types.Image, error) {
```

Comment:
This feels like a utils.go kind of function

Reply:
I am making this an exported function.  Remember I wanted a set of these to use internally and externally thereby reducing burden on the developer and code bloat.

Reply:
I am REALLY against exposing this, because it would expose ImageRef as well. We don't want to expose ImageRef, or any of the other refs, outside the API. I do not want to force people using the API to know what the difference between an Image, an ImageRef, and a StorageRef is.